addon/PlacesProvider.js

```diff
@@ -391,7 +405,8 @@ Links.prototype = {
                       WHERE hidden = 0 AND last_visit_date NOTNULL
                       AND moz_places.url NOT IN (${blockedURLs})
                       ORDER BY frecency, last_visit_date, moz_places.url DESC
-                    ) GROUP BY rev_host)
+                    ) GROUP BY rev_host
+                      ORDER BY frecency, last_visit_date, url DESC)
```

Comment:
As noted on irc, we'll probably end up implementing this type of custom combining/sorting logic in js so that the query is just a frecency-indexed `SELECT` to be fast. Having the unit tests will help make sure we maintain the desired logic while we rewrite.